File: crates/agent-protocol/src/app_game_adapter_execution_readiness.rs
use serde::{Deserialize, Serialize};

pub const APP_GAME_ADAPTER_EXECUTION_READINESS_READ_MODEL_ID: &str =
    "app-game-adapter-execution-readiness";
pub const APP_GAME_ADAPTER_EXECUTION_READINESS_CUSTODY_SUPPORTED_ADAPTER_RUNTIME_PROOF: &str =
    "supported-adapter-runtime-proof";
pub const APP_GAME_ADAPTER_EXECUTION_READINESS_STATUS_PARTIAL: &str =
    "app-game-adapter-execution-partial";
pub const APP_GAME_ADAPTER_PRODUCT_NATIVE_APP: &str = "native-app";
pub const APP_GAME_ADAPTER_PRODUCT_NATIVE_GAME: &str = "native-game";
pub const APP_GAME_ADAPTER_EXECUTION_ROW_ID_PREFIX: &str = "app-game-adapter-execution-";
pub const APP_GAME_ADAPTER_EXECUTION_STATE_PROVED_SCOPED: &str = "proved-scoped-execution";
pub const APP_GAME_ADAPTER_EXECUTION_STATE_MANUAL_REQUIRED: &str = "manual-required";
pub const APP_GAME_ADAPTER_EXECUTION_STATE_UNAVAILABLE: &str = "unavailable";
pub const APP_GAME_ADAPTER_EXECUTION_STATE_UNSUPPORTED: &str = "unsupported";
pub const APP_GAME_ADAPTER_EXECUTION_STATE_DEGRADED: &str = "degraded";
pub const APP_GAME_ADAPTER_EXECUTION_STATE_NOT_CLAIMED: &str = "not-claimed";
pub const APP_GAME_ADAPTER_EXECUTION_DECISION_ALLOWED: &str = "execution-allowed";
pub const APP_GAME_ADAPTER_EXECUTION_DECISION_BLOCKED: &str = "blocked-before-execution";
pub const APP_GAME_ADAPTER_HOST_CAPABILITY_AVAILABLE: &str = "available";
pub const APP_GAME_ADAPTER_HOST_CAPABILITY_NOT_DETECTED: &str = "not-detected";
pub const APP_GAME_ADAPTER_HOST_CAPABILITY_NOT_APPLICABLE: &str = "not-applicable";
pub const APP_GAME_PARENT_PLATFORM_MACOS: &str = "macos";
pub const APP_GAME_PARENT_PLATFORM_IOS: &str = "ios";

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGameAdapterExecutionReadinessRow {
    pub schema_version: u16,
    pub row_id: String,
    pub source_proof_entry_id: String,
    pub platform: String,
    pub product_meanings: Vec<String>,
    pub adapter_capability: String,
    pub adapter_execution_state: String,
    pub execution_decision: String,
    pub runtime_boundary: String,
    pub target_identity_state: String,
    pub rollback_reference_state: String,
    pub audit_reference_state: String,
    pub evidence_refs: Vec<String>,
    pub host_capability_state: String,
    pub host_capability_evidence_refs: Vec<String>,
    pub host_capability_probe_refs: Vec<String>,
    pub linked_proof_artifacts: Vec<String>,
    pub manual_proof_requirements: Vec<String>,
    pub claim_boundary: String,
    pub fallback_behavior: String,
    pub adapter_execution_claimed: bool,
    pub broad_installed_app_blocking_claimed: bool,
    pub child_device_delivery_claimed: bool,
    pub platform_enforcement_claimed: bool,
    pub provider_delivery_claimed: bool,
    pub private_diagnostics_claimed: bool,
    pub last_checked_at: String,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppGameAdapterExecutionReadinessReadModel {
    pub schema_version: u16,
    pub read_model_id: String,
    pub generated_at: String,
    pub source_read_model_ids: Vec<String>,
    pub custody_label: String,
    pub capability_status: String,
    pub returned: u64,
    pub execution_allowed_count: u64,
    pub blocked_before_execution_count: u64,
    pub adapter_execution_claimed_count: u64,
    pub host_capability_available_count: u64,
    pub host_capability_not_detected_count: u64,
    pub host_capability_not_applicable_count: u64,
    pub host_capability_probe_ref_count: u64,
    pub broad_installed_app_blocking_claimed: bool,
    pub child_device_delivery_claimed: bool,
    pub platform_enforcement_claimed: bool,
    pub provider_delivery_claimed: bool,
    pub private_diagnostics_claimed: bool,
    pub rows: Vec<AppGameAdapterExecutionReadinessRow>,
}

File: packages/app-game-domain/src/app-game-adapter-execution-readiness.ts
import { Schema } from '@ocentra-parent/schema-domain/effect';
import {
  AppGameAdapterExecutionReadinessReadModelSchema,
  AppGameAdapterExecutionReadinessReferenceSchema,
  AppGameAdapterExecutionReadinessRowSchema,
  type AppGameAdapterExecutionReadinessReadModel as AppGameAdapterExecutionReadinessReadModelShape,
  type AppGameAdapterExecutionReadinessReference,
  type AppGameAdapterExecutionReadinessRow,
  type AppGameAdapterExecutionState,
  type AppGameAdapterHostCapabilityState,
} from '@ocentra-parent/schema-domain/app-game-adapter-execution-readiness';
import { ParentContractSchemaVersion } from '@ocentra-parent/schema-domain/family-reference-primitives';
import {
  V08SupportedAdapterRuntimeProofReadModel,
  type V08SupportedAdapterRuntimeProofEntry,
  type V08SupportedAdapterRuntimeProofReadModel as V08SupportedAdapterRuntimeProofReadModelType,
} from '@ocentra-parent/schema-domain/v0-8-supported-adapter-runtime-proof';

const generatedAt = '2026-06-08T09:17:00.000Z';

const AppGameAdapterRuntimeProofEntryIds = new Set([
  'windows-app-game-owned-process-time-limit',
  'windows-broad-installed-app-blocking-manual-gate',
  'windows-broad-installed-app-artifact-status',
  'windows-adapter-permission-dependency-degraded',
  'linux-host-adapter-unavailable',
  'macos-host-adapter-unsupported',
  'android-mobile-control-manual-gate',
  'ios-mobile-control-manual-gate',
]);
const decodeAppGameAdapterExecutionReadinessReference = Schema.decodeUnknownSync(
  AppGameAdapterExecutionReadinessReferenceSchema
);

export const AppGameAdapterExecutionReadinessReadModel = buildAppGameAdapterExecutionReadinessReadModel(
  V08SupportedAdapterRuntimeProofReadModel
);

export function buildAppGameAdapterExecutionReadinessReadModel(
  sourceReadModel: V08SupportedAdapterRuntimeProofReadModelType
): AppGameAdapterExecutionReadinessReadModelShape {
  return AppGameAdapterExecutionReadinessReadModelSchema.parse({
    schemaVersion: ParentContractSchemaVersion.V0_6,
    readModelId: 'app-game-adapter-execution-readiness',
    generatedAt,
    sourceReadModelIds: [String(sourceReadModel.readModelId)],
    rows: sourceReadModel.entries
      .filter((entry: V08SupportedAdapterRuntimeProofEntry) =>
        AppGameAdapterRuntimeProofEntryIds.has(String(entry.proofEntryId))
      )
      .map(appGameAdapterExecutionReadinessRowFromEntry),
  });
}

function appGameAdapterExecutionReadinessRowFromEntry(
  entry: V08SupportedAdapterRuntimeProofEntry
): AppGameAdapterExecutionReadinessRow {
  const executionState = appGameAdapterExecutionState(entry);
  const executionAllowed = executionState === 'proved-scoped-execution';
  return AppGameAdapterExecutionReadinessRowSchema.parse({
    schemaVersion: ParentContractSchemaVersion.V0_6,
    rowId: `app-game-adapter-execution-${entry.proofEntryId}`,
    sourceProofEntryId: entry.proofEntryId,
    platform: entry.platform,
    productMeanings: ['native-app', 'native-game'],
    adapterCapability: entry.adapterCapability,
    adapterExecutionState: executionState,
    executionDecision: executionAllowed ? 'execution-allowed' : 'blocked-before-execution',
    runtimeBoundary: entry.runtimeBoundary,
    targetIdentityState: entry.targetIdentityState,
    rollbackReferenceState: entry.rollbackReferenceState,
    auditReferenceState: entry.auditReferenceState,
    evidenceRefs: entry.evidenceRefs,
    hostCapabilityState: appGameHostCapabilityState(entry),
    hostCapabilityEvidenceRefs: appGameHostCapabilityEvidenceRefs(entry),
    hostCapabilityProbeRefs: appGameHostCapabilityProbeRefs(entry),
    linkedProofArtifacts: entry.linkedProofArtifacts,
    manualProofRequirements: executionAllowed ? [] : appGameManualProofRequirements(entry),
    claimBoundary: appGameClaimBoundary(entry),
    fallbackBehavior: entry.fallbackBehavior,
    adapterExecutionClaimed: executionAllowed,
    broadInstalledAppBlockingClaimed: false,
    childDeviceDeliveryClaimed: false,
    platformEnforcementClaimed: false,
    providerDeliveryClaimed: false,
    privateDiagnosticsClaimed: false,
    lastCheckedAt: generatedAt,
  });
}

function appGameHostCapabilityProbeRefs(
  entry: V08SupportedAdapterRuntimeProofEntry
): readonly AppGameAdapterExecutionReadinessReference[] {
  if (entry.platform === 'macos' || entry.platform === 'ios') {
    return appGameReadinessReferences([]);
  }
  if (entry.platform === 'windows') {
    return appGameReadinessReferences(['windows-host-local-probe-ref']);
  }
  return appGameReadinessReferences([]);
}

function appGameHostCapabilityState(entry: V08SupportedAdapterRuntimeProofEntry): AppGameAdapterHostCapabilityState {
  if (entry.platform === 'windows') {
    return 'available';
  }
  if (entry.platform === 'macos' || entry.platform === 'ios') {
    return 'not-applicable';
  }
  return 'not-detected';
}

function appGameHostCapabilityEvidenceRefs(
  entry: V08SupportedAdapterRuntimeProofEntry
): readonly AppGameAdapterExecutionReadinessReference[] {
  if (entry.platform === 'windows') {
    return appGameReadinessReferences(['adapter-capability-state-ref']);
  }
  return appGameReadinessReferences([]);
}

function appGameAdapterExecutionState(entry: V08SupportedAdapterRuntimeProofEntry): AppGameAdapterExecutionState {
  if (entry.proofEntryId === 'windows-app-game-owned-process-time-limit') {
    return 'proved-scoped-execution';
  }
  if (entry.runtimeState === 'manual-required') return 'manual-required';
  if (entry.runtimeState === 'unavailable') return 'unavailable';
  if (entry.runtimeState === 'unsupported') return 'unsupported';
  if (entry.runtimeState === 'degraded') return 'degraded';
  return 'not-claimed';
}

function appGameManualProofRequirements(
  entry: V08SupportedAdapterRuntimeProofEntry
): readonly AppGameAdapterExecutionReadinessReference[] {
  if (entry.manualProofRequirements.length > 0) {
    return appGameReadinessReferences(entry.manualProofRequirements);
  }
  return appGameReadinessReferences(['app-game adapter execution proof requirement']);
}

function appGameClaimBoundary(entry: V08SupportedAdapterRuntimeProofEntry): string {
  if (entry.proofEntryId === 'windows-app-game-owned-process-time-limit') {
    return `${entry.claimBoundary} Adapter execution is claimable only for scoped owned-process time-limit rows.`;
  }
  return `${entry.claimBoundary} Adapter execution is blocked before runtime for this app/game row.`;
}

function appGameReadinessReferences(values: readonly string[]): readonly AppGameAdapterExecutionReadinessReference[] {
  return values.map((value) => decodeAppGameAdapterExecutionReadinessReference(value));
}

export function summarizeAppGameAdapterExecutionReadiness(
  readModel: AppGameAdapterExecutionReadinessReadModelShape
): Record<string, number> {
  return {
    rows: readModel.rows.length,
    executionAllowed: readModel.rows.filter((row) => row.executionDecision === 'execution-allowed').length,
    blockedBeforeExecution: readModel.rows.filter((row) => row.executionDecision === 'blocked-before-execution').length,
    adapterExecutionClaimed: readModel.rows.filter((row) => row.adapterExecutionClaimed).length,
    broadInstalledAppBlockingClaimed: readModel.rows.filter((row) => row.broadInstalledAppBlockingClaimed).length,
    childDeviceDeliveryClaimed: readModel.rows.filter((row) => row.childDeviceDeliveryClaimed).length,
    platformEnforcementClaimed: readModel.rows.filter((row) => row.platformEnforcementClaimed).length,
    providerDeliveryClaimed: readModel.rows.filter((row) => row.providerDeliveryClaimed).length,
    privateDiagnosticsClaimed: readModel.rows.filter((row) => row.privateDiagnosticsClaimed).length,
  };
}

export const decodeAppGameAdapterExecutionReadinessRow = Schema.decodeUnknownSync(
  AppGameAdapterExecutionReadinessRowSchema
);
export const decodeAppGameAdapterExecutionReadinessReadModel = Schema.decodeUnknownSync(
  AppGameAdapterExecutionReadinessReadModelSchema
);
